### PR TITLE
add gverb and glm as cmake external projects

### PR DESCRIPTION
**Pull request by birarda**

### CODE REVIEW

**Comment by birarda** [comment]
This is the beginning of saner dependency handling across all platforms - I will slowly move other dependencies to this flow if we do not run into snags with GLM + gverb, which are the first two I've replaced.

File: BUILD.md [hunk 1]
```diff
@@ -10,6 +9,12 @@
 * [Bullet Physics Engine](https://code.google.com/p/bullet/downloads/list) ~> 2.82
 * [Gverb](https://github.com/highfidelity/gverb/archive/master.zip) (direct download to latest version)
 
+#### CMake External Project Dependencies
+
+The following dependencies will be downloaded, built, linked and included automatically by CMake where we require them. The CMakeLists files that handle grabbing each of the following external dependencies can be found in the [cmake/externals folder](cmake/externals). The resulting downloads, source files and binaries will be placed in the `build` directory in each of the subfolders for each external project. These are not placed in your normal build tree when doing an out of source build so that they do not need to be re-downloaded and re-compiled every time the CMake build folder is cleared.
+
+* [glm](http://glm.g-truc.net/0.9.5/index.html) ~> 0.9.5.4
```
Comment (ZappoMan) [inline BUILD.md]: did you mean to include gverb here as well?

Comment (birarda) [inline BUILD.md]: I will now, thanks

File: BUILD_WIN.md [hunk 2]
```diff
@@ -129,12 +126,6 @@ Download the binary package: `glew-1.10.0-win32.zip`. Extract to %HIFI_LIB_DIR%\
...
 
-###GLM
-
-This package contains only headers, so there's nothing to add to the PATH.
-
-Be careful with glm. For the folder other libraries would normally call 'include', the folder containing the headers, glm opts to use 'glm'. You will have a glm folder nested inside the top-level glm folder.
-
 ###Gverb
 
 1. Go to https://github.com/highfidelity/gverb
```
Comment (ZappoMan) [inline BUILD_WIN.md]: did you mean to leave this in here? did you make gverb work as an automatic external?

Comment (birarda) [inline BUILD_WIN.md]: missed this - fixing

**Comment by birarda** [comment]
Investigating why ubuntu is failing to download GLM - win and OS X seem to have no problem grabbing both.

File: cmake/externals/gverb/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+if (ANDROID)
+  set(ANDROID_CMAKE_ARGS "-DCMAKE_TOOLCHAIN_FILE=${CMAKE_TOOLCHAIN_FILE}" "-DANDROID_NATIVE_API_LEVEL=19")
+endif ()
+
+include(ExternalProject)
+ExternalProject_Add(
+  ${EXTERNAL_NAME}
+  PREFIX ${EXTERNAL_NAME}
+  GIT_REPOSITORY https://github.com/birarda/gverb.git
```
Comment (ZappoMan) [inline cmake/externals/gverb/CMakeLists.txt]: hmmm... can you explain this? I'm guessing that gverb doesn't have an external path do download from? Or what? Are we going to have to make external git repos like this often? Also, I'm curious why is this in birarda and not something hifi? was that just a test? or did you intentionally not want it associated with hifi?

Comment (birarda) [inline cmake/externals/gverb/CMakeLists.txt]: We have a hifi repo - I didn't want to replace what is there until this is merged, at which point I'll switch to HF repo for the download link.

We have some small patches to gverb and I think it's also not easily findable - so this is a rare exception where we will need to host.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 48e7aa74330135185ffcd49a4f9bb2444487d0b2 available at: https://gist.github.com/6c398a2163d5501c7ace<br/><br/>Installers available for download at:<br>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4289.exe<br>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4289.dmg

**Comment by OmegaHeron** [comment]
How will this work if end user has a distribution's package system installed (since it applies) GLM?  Will it detect that being per-installed and use it, download your specific package using it or something else?  My concern is that we don't end up with clashing multiple installs of some of these dependencies making life on Linux more difficult vs less.

**Comment by birarda** [comment]
@OmegaHeron I plan on adding an option so you can disable us grabbing the external and pass a specific path or have it look in sys location, if you choose. Would that handle that case?

**Comment by OmegaHeron** [comment]
I think so as long as that's made clear - going forward I like the idea of fetching dependencies at level of HiFi's liking vs, possibly, far removed from packaged choices.  Just don't want to see the few who are already slugging it out with Linux find life complicated more so.

**Comment by ZappoMan** [comment]
@OmegaHeron - Note: One reason for making this change is that we've found cases where there may be bugs or conflicts in some versions of these dependencies... so if you aren't building with our official supported external dependencies, then we may not support addressing those bugs.

**Comment by birarda** [comment]
To specifically answer the question though @OmegaHeron - to guarantee version of package is correct at least in the short term I imagine we will prefer grabbing it with Cmake and containing it to the source dir (where not sticking it in the path where it could cause you issues with other things)

**Comment by OmegaHeron** [comment]
Don't get me wrong - I like this idea.  It's been a challenge to stay in sync 100% with you internal choices - especially for those who resist using anything but packages... this would help rectify much of that.  :)

File: cmake/externals/gverb/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+if (ANDROID)
+  set(ANDROID_CMAKE_ARGS "-DCMAKE_TOOLCHAIN_FILE=${CMAKE_TOOLCHAIN_FILE}" "-DANDROID_NATIVE_API_LEVEL=19")
+endif ()
+
+include(ExternalProject)
+ExternalProject_Add(
+  ${EXTERNAL_NAME}
+  PREFIX ${EXTERNAL_NAME}
+  GIT_REPOSITORY http://github.com/birarda/gverb.git
```
Comment (ZappoMan) [inline cmake/externals/gverb/CMakeLists.txt]: I put a comment here but it seems to have gotten eaten... why is this in the birarda repo? Should we add this to a hifi repo?

Comment (ZappoMan) [inline cmake/externals/gverb/CMakeLists.txt]: ok - I see your comment in email... strange that it got deleted.

**Comment by hifi-gustavo** [comment]
Build console output for commit: d337cadb106f8787ac40f1afe8b0c50fd023e3c4 available at: https://gist.github.com/2e6a1f6436d6a94a62ad<br/><br/>Installers available for download at:<br>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4289.exe<br>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4289.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 691cf2d893ba783a589615b996aef751eb192f78 available at: https://gist.github.com/b0031ad669d9530754d5<br/><br/>Installers available for download at:<br>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4289.exe<br>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4289.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 28fabcb7884670eb3c562de570c9ca68130431cf available at: https://gist.github.com/414e0ff0c7245083174b<br/><br/>Installers available for download at:<br>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4289.exe<br>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4289.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 018b8497bddee1b64a8196975e492d0131221ff5 available at: https://gist.github.com/e7724315c5763cfa28be<br/><br/>Installers available for download at:<br>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4289.exe<br>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4289.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9683ed92e321cdec5b7607a03f0bf08dd1753fd5 available at: https://gist.github.com/56c3f22ac7fde78ce2e7<br/><br/>Installers available for download at:<br>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4289.exe<br>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4289.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: a1bf06f004406a3800c76e1d1e65d848993cbd5d available at: https://gist.github.com/92cbe8b89d6ec7bcff9b<br/><br/>Installers available for download at:<br>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4289.exe<br>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4289.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: feaad377514864c1b5347b532566923a66d32173 available at: https://gist.github.com/a456098d28eaa648b4bc<br/><br/>Installers available for download at:<br>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4289.exe<br>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4289.dmg

File: cmake/macros/AddDependencyExternalProject.cmake [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+      else ()
+        set(_PROJ_BINARY_DIR ${EXTERNAL_PROJECT_DIR}/${_PROJ_NAME}/build)
+      endif ()
+    
+      add_subdirectory(${EXTERNAL_PROJECT_DIR}/${_PROJ_NAME} ${_PROJ_BINARY_DIR})
+    endif ()
+  
+    add_dependencies(${TARGET_NAME} ${_PROJ_NAME})
+    
+  endif ()
```
Comment (AndrewMeadows) [inline cmake/macros/AddDependencyExternalProject.cmake]: I don't understand line 16.  Why is it: GET_${FOO} OR GET_${FOO} ?

Shouldn't it be: GET_${FOO} OR GET_${BAR} ?

Comment (birarda) [inline cmake/macros/AddDependencyExternalProject.cmake]: it is `if (NOT DEFINED GET_${_PROJ_NAME_UPPER} OR GET_${_PROJ_NAME_UPPER})`

we want to default it to true if it isn't passed

**Comment by hifi-gustavo** [comment]
Build console output for commit: d299f015d1f7e2e549cb14b869e34b5a4fe6a146 available at: https://gist.github.com/611ec25d58c2c1127b23<br/><br/>Installers available for download at:<br>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4289.exe<br>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4289.dmg

**Pull request merged**